components/hab/src/command/pkg/binlink.rs

```diff
@@ -135,89 +141,159 @@ pub fn binlink_all_in_pkg(ui: &mut UI,
     Ok(())
 }
 
-fn is_dest_on_path<T: AsRef<Path>>(dest_dir: T) -> bool {
+fn is_dest_on_path(dest_dir: &Path) -> bool {
     if let Some(val) = env::var_os("PATH") {
-        env::split_paths(&val).any(|p| p == dest_dir.as_ref())
+        env::split_paths(&val).any(|p| p == dest_dir)
     } else {
         false
     }
 }
 
+#[allow(dead_code)]
 struct Binlink {
-    dest: PathBuf,
-    src:  PathBuf,
+    dest:                PathBuf,
+    src:                 PathBuf,
+    default_interpreter: PathBuf,
 }
 
-#[cfg(not(target_os = "windows"))]
 impl Binlink {
-    pub fn new<T: AsRef<Path>>(src: T, dest_dir: T) -> Result<Self> {
-        let bin_name = match src.as_ref().file_name() {
-            Some(name) => name,
-            None => return Err(Error::CannotParseBinlinkSource(src.as_ref().to_path_buf())),
-        };
-
-        Ok(Self { dest: dest_dir.as_ref().join(bin_name),
-                  src:  src.as_ref().to_path_buf(), })
+    pub fn new(src: &Path, dest_dir: &Path, default_interpreter: &Path) -> Result<Self> {
+        Ok(Self { dest:                Self::binstub_path(&src, dest_dir)?,
+                  src:                 src.to_path_buf(),
+                  default_interpreter: default_interpreter.to_path_buf(), })
     }
 
-    pub fn from_file<T: AsRef<Path>>(dest: T) -> Result<Self> {
-        Ok(Binlink { dest: dest.as_ref().to_path_buf(),
-                     src:  fs::read_link(&dest)?, })
+    pub fn from_file(path: &Path) -> Result<Self> {
+        use toml::Value::Table;
+
+        // its possible the link could already exist as a symlink either
+        // because a previous version of habitat did the binlinking or
+        // the symlink created outside of habitat
+        match fs::read_link(&path) {
+            Ok(lnk) => {
+                Ok(Binlink { dest:                path.to_path_buf(),
+                             src:                 lnk,
+                             default_interpreter: PathBuf::from(DEFAULT_INTERPRETER), })
+            }
+            Err(_) => {
+                let file = File::open(path)?;
+                for line in BufReader::new(file).lines() {
+                    let ln = line?;
+                    if ln.to_uppercase().starts_with(COMMENT_MARKER) {
+                        let (_, rest) = ln.split_at(COMMENT_MARKER.len());
+                        if let Ok(Table(toml_exp)) = rest.parse() {
```

Comment:
The cascade of `if let`s may be more readable re-cast with an FP approach:
```rust
                rest.parse()
                    .map(|Table(toml_exp)| toml_exp.get("source"))
                    .map(toml::value::Value::as_str)
                    .map(|val| ...);
```
And then we might be able to reuse `Self::from_link` at the bottom.

Reply:
I poked at this a bit, and here's an approach that will actually compile:
```rust
                if let Some(source) = Self::get_source_from_toml(rest) {
                    return Ok(Self::with_default_interpreter(source.into(), path.into()));
                }
```
with
```rust
    fn with_default_interpreter(src: PathBuf, dest: PathBuf) -> Self {
        Self { dest,
               src,
               default_interpreter: PathBuf::from(DEFAULT_INTERPRETER) }
    }

    fn get_source_from_toml(toml: &str) -> Option<String> {
        toml.parse()
            .ok()
            .as_ref()
            .and_then(toml::value::Value::as_table)
            .and_then(|toml_table| toml_table.get("source"))
            .and_then(toml::value::Value::as_str)
            .map(String::from)
    }

```